File: 16-3sum-closest/3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        
        # focus on minimizing the difference between the target and the sum.

        diff = float('inf')
        nums.sort()
        for i, a in enumerate(nums):
            l, r = i+1, len(nums)-1
            while(l<r):
                sum = a + nums[l] + nums[r]

                if abs(target-sum) <  abs(diff):
                    diff = abs(target-sum)

                    # keep track of this sum to return it as result
                    res = sum

                if sum > target:
                    r-=1
                else:
                    l+=1
            
        return res

File: 16-3sum-closest/3sum-closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {

        int diff = Integer.MAX_VALUE, sum=0;

        Arrays.sort(nums);
        
        for(int i =0; i<nums.length && diff!=0; i++){
            int l = i+1;
            int r = nums.length-1;

            while(l<r){

                sum = nums[i] + nums[l] + nums[r];

                if(Math.abs(target - sum) < Math.abs(diff)){
                    diff = target - sum;
                }

                if(sum < target){
                    l++;
                }
                else{
                    r--;
                }
            }
        }
        
        return target-diff;
    }
}